File: mixer-behringer/src/main/kotlin/org/openmultitrack/mixer/behringer/MixingStationIcons.kt
package org.openmultitrack.mixer.behringer

/**
 * Glyphs for mixer scribble icon ids (1–74).
 * Numbering matches the X32 / X-Air / Mixing Station / FLOW 8 icon id scheme.
 *
 * Rendered as Unicode emoji or short desk-style abbreviations — not copied mixer artwork.
 * Abbrevs are used where emoji are misleading or identical (drums, jacks, guitars).
 */
object MixingStationIcons {
    const val MAX_ID = 74

    const val ELECTRIC_BASS = 17
    const val ACOUSTIC_GUITAR = 23
    const val VIOLIN = 39
    const val HANDHELD_MIC = 50
    const val TAPE = 60
    const val PC = 62
    const val SPEAKER_RIGHT = 64
    const val SPEAKER_LEFT = 65

    enum class GlyphStyle {
        EMOJI,
        /** Short label (BD, XLR, …) rendered in a smaller mono-style strip cell. */
        ABBREV,
    }

    data class Glyph(
        val text: String,
        val style: GlyphStyle = GlyphStyle.EMOJI,
    )

    private val EMOJI = arrayOf(
        "", // 0 unused
        "", // 1 blank
        "🥁", // 2 kick-back — fallback; display() prefers BD
        "🥁", // 3 kick-front
        "🪘", // 4 snare-top
        "🪘", // 5 snare-bottom
        "🥁", // 6 tom-high
        "🥁", // 7 tom-medium
        "🥁", // 8 floor tom
        "🎩", // 9 hi-hat — fallback; display() prefers HH
        "🔔", // 10 crash
        "🥁", // 11 drum kit
        "🔔", // 12 cowbell
        "🪘", // 13 bongos
        "🪘", // 14 congas
        "🎵", // 15 tambourine
        "🎵", // 16 vibraphone
        "🎸", // 17 electric bass
        "🎸", // 18 acoustic bass
        "🎸", // 19 contrabass
        "🎸", // 20 les paul
        "🎸", // 21 ibanez
        "🎸", // 22 washburn
        "🎸", // 23 acoustic guitar
        "🔊", // 24 bass amp
        "🔊", // 25 guitar amp
        "🔊", // 26 amp cabinet
        "🎹", // 27 piano
        "🎹", // 28 organ
        "🎹", // 29 harpsichord
        "🎹", // 30 keyboard
        "🎹", // 31 synthesizer 1
        "🎹", // 32 synthesizer 2
        "🎹", // 33 synthesizer 3
        "🎹", // 34 keytar
        "🎺", // 35 trumpet
        "🎺", // 36 trombone
        "🎷", // 37 saxophone
        "🎷", // 38 clarinet
        "🎻", // 39 violin
        "🎻", // 40 cello
        "🎤", // 41 male vocal
        "🎤", // 42 female vocal
        "👥", // 43 choir
        "✋", // 44 hand sign
        "🗣", // 45 talk a
        "🗣", // 46 talk b
        "🎙", // 47 large diaphragm mic
        "🎙", // 48 condenser mic left
        "🎙", // 49 condenser mic right
        "🎤", // 50 handheld mic (SM58)
        "🎤", // 51 wireless mic
        "🎤", // 52 podium mic
        "🎧", // 53 headset / in-ear
        "🔌", // 54 xlr — fallback; display() prefers XLR
        "🔌", // 55 trs
        "🔌", // 56 trs left
        "🔌", // 57 trs right
        "🔌", // 58 rca left
        "🔌", // 59 rca right
        "📼", // 60 tape
        "✨", // 61 fx
        "💻", // 62 computer
        "🔊", // 63 wedge
        "🔈", // 64 speaker right
        "🔉", // 65 speaker left
        "🔊", // 66 speaker array
        "🔊", // 67 speaker on pole
        "🎛", // 68 amp rack
        "🎛", // 69 controls
        "🎚", // 70 fader
        "🔀", // 71 mix bus
        "🔀", // 72 matrix
        "🔀", // 73 routing
        "😊", // 74 smiley
    )

    /** Desk-style abbreviations for ids where emoji are wrong, ambiguous, or identical. */
    private val ABBREV = mapOf(
        2 to "BD",
        3 to "BD",
        4 to "SN",
        5 to "SN",
        6 to "T1",
        7 to "T2",
        8 to "FT",
        9 to "HH",
        10 to "CR",
        11 to "DK",
        12 to "CB",
        17 to "EB",
        18 to "AB",
        19 to "UB",
        20 to "EG",
        21 to "G2",
        22 to "G3",
        23 to "AG",
        54 to "XLR",
        55 to "TRS",
        56 to "T-L",
        57 to "T-R",
        58 to "C-L",
        59 to "C-R",
    )

    fun display(iconId: Int?): Glyph? {
        if (iconId == null || iconId !in 1..MAX_ID) return null
        ABBREV[iconId]?.let { return Glyph(it, GlyphStyle.ABBREV) }
        return EMOJI.getOrNull(iconId)?.takeIf { it.isNotEmpty() }?.let { Glyph(it) }
    }

    /** Legacy emoji-only lookup (abbrev icons return null). */
    fun emoji(iconId: Int?): String? {
        val glyph = display(iconId) ?: return null
        return if (glyph.style == GlyphStyle.EMOJI) glyph.text else null
    }

    /** Text for strip rendering — emoji or abbreviation. */
    fun stripText(iconId: Int?): String? = display(iconId)?.text
}

File: docs/flow8-reverse-engineering/tools/mixing_station_icons.py
"""Mixing Station / X32 scribble icon ids (1–74). Mirrors MixingStationIcons.kt."""

from __future__ import annotations

ICON_MIN = 1
ICON_MAX = 74

# Named constants (MixingStationIcons.kt)
HANDHELD_MIC = 50
ELECTRIC_BASS = 17
ACOUSTIC_GUITAR = 23
VIOLIN = 39
TAPE = 60
PC = 62
SPEAKER_RIGHT = 64
SPEAKER_LEFT = 65

ICON_CONSTANTS: dict[int, str] = {
    17: "ELECTRIC_BASS",
    23: "ACOUSTIC_GUITAR",
    39: "VIOLIN",
    50: "HANDHELD_MIC",
    60: "TAPE",
    62: "PC",
    64: "SPEAKER_RIGHT",
    65: "SPEAKER_LEFT",
}

# X32 / Mixing Station icon labels (behringer-icons / forum list)
ICON_LABELS: list[str] = [
    "",  # 0 unused
    "blank",
    "kick-back",
    "kick-front",
    "snare-top",
    "snare-bottom",
    "tom-high",
    "tom-medium",
    "floor-tom",
    "hi-hat",
    "crash",
    "drum-kit",
    "cowbell",
    "bongos",
    "congas",
    "tambourine",
    "vibraphone",
    "electric-bass",
    "acoustic-bass",
    "contrabass",
    "les-paul",
    "ibanez",
    "washburn",
    "acoustic-guitar",
    "bass-amp",
    "guitar-amp",
    "amp-cabinet",
    "piano",
    "organ",
    "harpsichord",
    "keyboard",
    "synthesizer-1",
    "synthesizer-2",
    "synthesizer-3",
    "keytar",
    "trumpet",
    "trombone",
    "saxophone",
    "clarinet",
    "violin",
    "cello",
    "male-vocal",
    "female-vocal",
    "choir",
    "hand-sign",
    "talk-a",
    "talk-b",
    "large-diaphragm-mic",
    "condenser-mic-left",
    "condenser-mic-right",
    "handheld-mic",
    "wireless-mic",
    "podium-mic",
    "headset-mic",
    "xlr",
    "trs",
    "trs-left",
    "trs-right",
    "rca-left",
    "rca-right",
    "tape",
    "fx",
    "computer",
    "wedge",
    "speaker-right",
    "speaker-left",
    "speaker-array",
    "speaker-on-pole",
    "amp-rack",
    "controls",
    "fader",
    "mix-bus",
    "matrix",
    "routing",
    "smiley",
]

ICON_EMOJI: list[str] = [
    "",
    "",
    "🥁", "🥁", "🪘", "🪘", "🥁", "🥁", "🥁", "🎩", "🔔",
    "🥁", "🔔", "🪘", "🪘", "🎵", "🎵", "🎸", "🎸", "🎸", "🎸",
    "🎸", "🎸", "🎸", "🔊", "🔊", "🔊", "🎹", "🎹", "🎹", "🎹",
    "🎹", "🎹", "🎹", "🎹", "🎺", "🎺", "🎷", "🎷", "🎻", "🎻",
    "🎤", "🎤", "👥", "✋", "🗣", "🗣", "🎙", "🎙", "🎙", "🎤",
    "🎤", "🎤", "🎧", "🔌", "🔌", "🔌", "🔌", "🔌", "🔌", "📼",
    "✨", "💻", "🔊", "🔈", "🔉", "🔊", "🔊", "🎛", "🎛", "🎚",
    "🔀", "🔀", "🔀", "😊",
]


def icon_label(icon_id: int | None) -> str | None:
    if icon_id is None or not (ICON_MIN <= icon_id <= ICON_MAX):
        return None
    return ICON_LABELS[icon_id] or None


def icon_constant(icon_id: int | None) -> str | None:
    if icon_id is None:
        return None
    return ICON_CONSTANTS.get(icon_id)


def icon_emoji(icon_id: int | None) -> str | None:
    if icon_id is None or not (ICON_MIN <= icon_id <= ICON_MAX):
        return None
    ch = ICON_EMOJI[icon_id]
    return ch or None


def format_icon(icon_id: int | None, flow_label: str | None = None) -> str:
    """Human-readable icon field for script output."""
    if icon_id is None:
        return "—"
    const = icon_constant(icon_id)
    label = icon_label(icon_id)
    emoji = icon_emoji(icon_id)
    parts = [f"{icon_id:02d}"]
    if flow_label:
        parts.append(flow_label)
    elif const:
        parts.append(const)
    elif label:
        parts.append(label)
    if emoji:
        parts.append(emoji)
    return " ".join(parts)
